File: csharp/948.bag-of-tokens.cs
/*
 * @lc app=leetcode id=948 lang=csharp
 *
 * [948] Bag of Tokens
 */

using System;

// @lc code=start
public class Solution {
    public int BagOfTokensScore(int[] tokens, int P) {
        int ret = 0, curr = 0, l = 0, r = tokens.Length - 1;
        Array.Sort(tokens);
        while (l <= r) {
            while (l <= r && tokens[l] <= P) {
                P -= tokens[l++];
                ++curr;
                ret = Math.Max(ret, curr);
            }
            if (curr == 0 || l > r) {
                break;
            }
            P += tokens[r--];
            --curr;
        }
        return ret;
    }
}
// @lc code=end

File: java/948.bag-of-tokens.java
/*
 * @lc app=leetcode id=948 lang=java
 *
 * [948] Bag of Tokens
 */

import java.util.Arrays;

// @lc code=start

class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        int n = tokens.length, l = 0, r = n - 1, ret = 0, s = 0, p = power;
        int[] sorted = Arrays.stream(tokens).sorted().toArray();
        while (l <= r) {
            if (p >= sorted[l]) {
                p -= sorted[l++];
                ++s;
            } else {
                if (s == 0) {
                    break;
                }
                p += sorted[r--];
                --s;
            }
            ret = Math.max(ret, s);
        }
        return ret;
    }
}
// @lc code=end

